sources/api/shibaken/src/main.rs

```diff
@@ -0,0 +1,278 @@
+fn fetch_public_keys_from_imds() -> Result<Vec<String>> {
+    info!("Fetching IMDS session token");
+    let client = Client::new();
+    let imds_session_token = fetch_imds_session_token(&client)?;
+
+    info!("Fetching list of available public keys from IMDS");
+    // Returns a list of available public keys as '0=my-public-key'
+    let public_key_list =
+        fetch_data_from_imds(&client, IMDS_PUBLIC_KEY_BASE_URL, &imds_session_token)?;
+    debug!("available public keys '{}'", &public_key_list);
+
+    info!("Generating endpoint urls to fetch text of available public keys");
+    let public_key_endpoints = build_public_key_endpoints_from_imds(&public_key_list)?;
+
+    info!("Fetching public keys from IMDS");
+    let mut public_keys = Vec::new();
+    for endpoint in public_key_endpoints.iter() {
+        let public_key_text = fetch_data_from_imds(&client, endpoint, &imds_session_token)?;
+        // Simple check to see if the text is probably an ssh key.
+        if &public_key_text[0..3] == "ssh" {
+            debug!("{}", &public_key_text.trim_end());
+            public_keys.push(public_key_text.trim_end().to_string())
+        } else {
+            warn!(
+                "'{}' does not appear to be a valid key. Skipping...",
+                &public_key_text.trim_end()
+            );
+            continue;
+        }
+    }
+    ensure!(!public_keys.is_empty(), error::NoValidImdsKeys);
+
+    Ok(public_keys)
+}
+
+/// Returns a list of endpoint urls strings for the public keys in IMDS. Since IMDS returns the list
+/// of available public keys as '0=my-public-key', we need to strip the index from the list and
+/// embed it into the endpoint url.
+fn build_public_key_endpoints_from_imds(public_key_list: &str) -> Result<Vec<String>> {
+    let mut public_key_endpoints = Vec::new();
+    for available_key in public_key_list.lines() {
+        let f: Vec<&str> = available_key.split('=').collect();
+        // If f[0] isn't a number, then it isn't a valid index.
+        if f[0].parse::<i32>().is_ok() {
+            let public_key_endpoint = format!("{}{}/openssh-key", IMDS_PUBLIC_KEY_BASE_URL, f[0]);
+            public_key_endpoints.push(public_key_endpoint);
+        } else {
+            warn!(
+                "'{}' does not appear to be a valid index. Skipping...",
+                &f[0]
+            );
+            continue;
+        }
+    }
+    ensure!(
+        !public_key_endpoints.is_empty(),
+        error::NoValidImdsKeyEndpoints
+    );
+    Ok(public_key_endpoints)
+}
+
+/// Store the args we receive on the command line.
+struct Args {
+    log_level: LevelFilter,
+}
+
+/// Print a usage message in the event a bad arg is passed.
+fn usage() -> ! {
+    let program_name = env::args().next().unwrap_or_else(|| "program".to_string());
+    eprintln!(
+        r"Usage: {}
+            [ --log-level trace|debug|info|warn|error ]",
+        program_name
+    );
+    process::exit(2);
+}
+
```

Comment:
@webern wants this handled differently to simplify testing.

Reply:
I'll go ahead and mirror the logic in https://github.com/bottlerocket-os/bottlerocket/pull/1317

Reply:
Can we say _why_ we want it handled differently or how it simplifies testing?  This new style using snafu and pushing the work into `main` seems more complicated.

Reply:
My understanding is this makes it so only `main` can exit the program to simplify testing. I think it would make sense for us to discuss as a team what our standard approach should be as there are pros/cons to both styles.